snuba/optimize.py

```diff
@@ -145,72 +169,141 @@ def _subdivide_parts(
-                        ignore_cutoff,
-                        clickhouse_host,
-                    ),
-                )
+        threaded_cutoff_time = (
+            last_midnight + settings.PARALLEL_OPTIMIZE_JOB_CUTOFF_TIME
+        )
+        logger.info(f"Threaded cutoff time: {threaded_cutoff_time}")
+
+    divided_parts = _subdivide_parts(parts, parallel)
+    threads: MutableSequence[threading.Thread] = []
+    for i in range(0, parallel):
+        threads.append(
+            threading.Thread(
+                target=optimize_partitions,
+                args=(
+                    clickhouse,
+                    database,
+                    table,
+                    divided_parts[i],
+                    threaded_cutoff_time,
+                    optimize_partition_tracker,
+                    clickhouse_host,
+                ),
             )
+        )
 
-            threads[i].start()
+        threads[i].start()
 
-        for i in range(0, parallel):
+    # Wait for all threads to finish. Any thread can raise JobTimeoutException. Its an indication that not all
+    # partitions could be optimized in the given amount of threaded_cutoff_time.
+    for i in range(0, parallel):
+        try:
             threads[i].join()
+        except JobTimeoutException:
+            pass
+
+    if not optimize_partition_tracker:
+        return
+
+    # Find out partitions which couldn't be optimized because of threaded cutoff time.
+    partitions = optimize_partition_tracker.get_completed_partitions()
+    remaining_parts = (
+        [part for part in parts if part.name not in partitions] if partitions else parts
+    )
+
+    if len(remaining_parts):
+        # Call optimize_partitions with the original cutoff_time now since running optimizations
+        # serially is allowed for a larger duration.
+        optimize_partitions(
+            clickhouse=clickhouse,
+            database=database,
+            table=table,
+            parts=remaining_parts,
+            cutoff_time=cutoff_time,
+            optimize_partition_tracker=optimize_partition_tracker,
+            clickhouse_host=clickhouse_host,
+        )
+
+
+def optimize_partition_runner(
+    clickhouse: ClickhousePool,
+    database: str,
+    table: str,
+    parts: Sequence[util.Part],
+    parallel: int,
+    cutoff_time: Optional[datetime],
+    optimize_partition_tracker: Optional[OptimizedPartitionTracker],
+    clickhouse_host: Optional[str] = None,
+) -> None:
+    """
+    Optimize partition runner decides whether to run the threaded version of
+    optimization or the non-threaded version.
+    """
+    if parallel <= 1:
+        optimize_partitions(
+            clickhouse=clickhouse,
+            database=database,
+            table=table,
+            parts=parts,
+            cutoff_time=cutoff_time,
+            optimize_partition_tracker=optimize_partition_tracker,
+            clickhouse_host=clickhouse_host,
+        )
```

Comment:
nit: I don;t think you are gaining much by having this two code paths handled separately. 
The job takes hours to complete, the threading overhead is none basically on that scale, why not just using the threaded version everywhere with 1 thread if you do not want parallelism?
So we will have only one code path to think about instead of two.

Reply:
Done